File: acrarium/src/main/java/com/faendir/acra/ui/view/app/tabs/admincards/DangerCard.java
/*
 * (C) Copyright 2019 Lukas Morawietz (https://github.com/F43nd1r)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.faendir.acra.ui.view.app.tabs.admincards;

import com.faendir.acra.i18n.Messages;
import com.faendir.acra.model.App;
import com.faendir.acra.service.DataService;
import com.faendir.acra.ui.component.*;
import com.faendir.acra.ui.component.dialog.FluentDialog;
import com.faendir.acra.ui.view.Overview;
import com.vaadin.flow.component.UI;
import com.vaadin.flow.component.combobox.ComboBox;
import com.vaadin.flow.component.textfield.NumberField;
import com.vaadin.flow.spring.annotation.SpringComponent;
import com.vaadin.flow.spring.annotation.UIScope;

import static com.faendir.acra.model.QReport.report;

@UIScope
@SpringComponent
public class DangerCard extends AdminCard {
    public DangerCard(DataService dataService) {
        super(dataService);
        setHeader(Translatable.createLabel(Messages.DANGER_ZONE));
        enableDivider();
        setHeaderColor("var(--lumo-error-contrast-color)", "var(--lumo-error-color)");
    }

    @Override
    public void init(App app) {
        removeContent();
        Box configBox = new Box(Translatable.createLabel(Messages.NEW_ACRA_CONFIG), Translatable.createLabel(Messages.NEW_ACRA_CONFIG_DETAILS),
                Translatable.createButton(e -> new FluentDialog().addText(Messages.NEW_ACRA_CONFIG_CONFIRM)
                        .addConfirmButtons(popup -> new FluentDialog().addComponent(new ConfigurationLabel(getDataService().recreateReporterUser(app))).addCloseButton().show())
                        .show(), Messages.CREATE));
        Box matchingBox = new Box(Translatable.createLabel(Messages.NEW_BUG_CONFIG), Translatable.createLabel(Messages.NEW_BUG_CONFIG_DETAILS),
                Translatable.createButton(e -> {
                    App.Configuration configuration = app.getConfiguration();
                    Translatable.ValidatedValue<RangeField, ?, Double> score = Translatable.createRangeField(Messages.SCORE).with(it -> {
                        it.setMin(0);
                        it.setMax(100);
                        it.setValue((double) configuration.getMinScore());
                    });
                    new FluentDialog().addComponent(score)
                            .addText(Messages.NEW_BUG_CONFIG_CONFIRM)
                            .addConfirmButtons(p -> getDataService().changeConfiguration(app, new App.Configuration(score.getValue().intValue())))
                            .show();
                }, Messages.CONFIGURE));
        Box purgeAgeBox = new Box(Translatable.createLabel(Messages.PURGE_OLD), Translatable.createLabel(Messages.PURGE_OLD_DETAILS), Translatable.createButton(e -> {
            Translatable.ValidatedValue<NumberField, ?, Double> age = Translatable.createNumberField(30d, Messages.REPORTS_OLDER_THAN1).with(it -> {
                        it.setStep(1d);
                        it.setMin(1d);
                        it.setPreventInvalidInput(true);
                        it.setHasControls(true);
                        it.setWidthFull();
                        it.setSuffixComponent(Translatable.createLabel(Messages.REPORTS_OLDER_THAN2));
                    }
            );
            new FluentDialog().addComponent(age)
                    .setTitle(Messages.PURGE)
                    .addConfirmButtons(popup -> {
                        getDataService().deleteReportsOlderThanDays(app, age.getValue().intValue());
                    }).show();
        }, Messages.PURGE));
        Box purgeVersionBox = new Box(Translatable.createLabel(Messages.PURGE_VERSION), Translatable.createLabel(Messages.PURGE_VERSION_DETAILS), Translatable.createButton(e -> {
            Translatable.ValidatedValue<ComboBox<Integer>, ?, Integer> versionBox = Translatable.createComboBox(getDataService().getFromReports(app, null, report.stacktrace.version.code), Messages.REPORTS_BEFORE_VERSION);
            new FluentDialog().addComponent(versionBox)
                    .setTitle(Messages.PURGE)
                    .addConfirmButtons(popup -> {
                        if (versionBox.getValue() != null) {
                            getDataService().deleteReportsBeforeVersion(app, versionBox.getValue());
                        }
                    }).show();
        }, Messages.PURGE));
        Box deleteBox = new Box(Translatable.createLabel(Messages.DELETE_APP), Translatable.createLabel(Messages.DELETE_APP_DETAILS), Translatable.createButton(e ->
                new FluentDialog().addText(Messages.DELETE_APP_CONFIRM).addConfirmButtons(popup -> {
                    getDataService().delete(app);
                    UI.getCurrent().navigate(Overview.class);
                }).show(), Messages.DELETE));
        add(configBox, matchingBox, purgeAgeBox, purgeVersionBox, deleteBox);
    }
}

File: acrarium/src/main/kotlin/com/faendir/acra/ui/view/app/tabs/admincards/DangerCard.kt
/*
 * (C) Copyright 2019 Lukas Morawietz (https://github.com/F43nd1r)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.faendir.acra.ui.view.app.tabs.admincards

import com.faendir.acra.i18n.Messages
import com.faendir.acra.jooq.generated.Tables.REPORT
import com.faendir.acra.navigation.RouteParams
import com.faendir.acra.navigation.View
import com.faendir.acra.persistence.app.AppRepository
import com.faendir.acra.persistence.report.ReportRepository
import com.faendir.acra.persistence.user.Permission
import com.faendir.acra.security.RequiresPermission
import com.faendir.acra.ui.component.AdminCard
import com.faendir.acra.ui.component.Translatable
import com.faendir.acra.ui.component.dialog.closeButton
import com.faendir.acra.ui.component.dialog.confirmButtons
import com.faendir.acra.ui.component.dialog.showFluentDialog
import com.faendir.acra.ui.ext.box
import com.faendir.acra.ui.ext.comboBox
import com.faendir.acra.ui.ext.configurationLabel
import com.faendir.acra.ui.ext.content
import com.faendir.acra.ui.ext.translatableNumberField
import com.faendir.acra.ui.ext.translatableText
import com.faendir.acra.ui.view.Overview
import com.vaadin.flow.component.UI
import java.time.Instant
import java.time.temporal.ChronoUnit

@View
@RequiresPermission(Permission.Level.ADMIN)
class DangerCard(
    appRepository: AppRepository,
    reportRepository: ReportRepository,
    routeParams: RouteParams,
) : AdminCard() {
    private val appId = routeParams.appId()

    init {
        content {
            setHeader(Translatable.createLabel(Messages.DANGER_ZONE))
            setHeaderColor("var(--lumo-error-contrast-color)", "var(--lumo-error-color)")
            dividerEnabled = true
            box(Messages.NEW_ACRA_CONFIG, Messages.NEW_ACRA_CONFIG_DETAILS, Messages.CREATE) {
                showFluentDialog {
                    translatableText(Messages.NEW_ACRA_CONFIG_CONFIRM)
                    confirmButtons {
                        showFluentDialog {
                            configurationLabel(appRepository.recreateReporter(appId))
                            closeButton()
                        }
                    }
                }
            }
            box(Messages.PURGE_OLD, Messages.PURGE_OLD_DETAILS, Messages.PURGE) {
                showFluentDialog {
                    header(Messages.PURGE)
                    val age = translatableNumberField(Messages.REPORTS_OLDER_THAN1) {
                        value = 30.0
                        step = 1.0
                        min = 1.0
                        isStepButtonsVisible = true
                        setWidthFull()
                        suffixComponent = Translatable.createLabel(Messages.REPORTS_OLDER_THAN2)
                    }
                    confirmButtons {
                        reportRepository.deleteBefore(appId, Instant.now().minus(age.getValue().toLong(), ChronoUnit.DAYS))
                    }
                }
            }
            box(Messages.PURGE_VERSION, Messages.PURGE_VERSION_DETAILS, Messages.PURGE) {
                showFluentDialog {
                    header(Messages.PURGE)
                    val versionBox = comboBox(reportRepository.get(appId, REPORT.VERSION_CODE), Messages.REPORTS_BEFORE_VERSION)
                    confirmButtons {
                        if (versionBox.value != null) {
                            reportRepository.deleteBefore(appId, versionBox.value!!)
                        }
                    }
                }
            }
            box(Messages.DELETE_APP, Messages.DELETE_APP_DETAILS, Messages.DELETE) {
                showFluentDialog {
                    translatableText(Messages.DELETE_APP_CONFIRM)
                    confirmButtons {
                        appRepository.delete(appId)
                        UI.getCurrent().navigate(Overview::class.java)
                    }
                }
            }
        }
    }
}